lisa/analysis/pixel6.py

```diff
@@ -0,0 +1,89 @@
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import pandas as pd
+import holoviews as hv
+from holoviews import opts
+
+from lisa.datautils import df_add_delta
+from lisa.analysis.base import TraceAnalysisBase
+from lisa.trace import requires_events
+from lisa.notebook import plot_signal
+
+
+class Pixel6Analysis(TraceAnalysisBase):
+    """
+    Support for Pixel 6-specific analysis data
+
+    :param trace: input Trace object
+    :type trace: :class:`lisa.trace.Trace`
+    """
+
+    name = 'pixel6'
+
+    EMETER_CHAN_NAMES = {
+        'S4M_VDD_CPUCL0': 'little',
+        'S3M_VDD_CPUCL1': 'mid',
+        'S2M_VDD_CPUCL2': 'big',
+    }
+
+###############################################################################
+# DataFrame Getter Methods
+###############################################################################
+    @TraceAnalysisBase.cache
+    @requires_events('pixel6_emeter')
+    def df_power_meter(self):
+        """
+        Get the power meter readings across the trace.
+
+        :retuns: A :class:`pandas.DataFrame` with:
+            * A ``channel`` column (description of the power usage channel)
+            * A ``value`` column (average power usage in mW since the last measurement)
+        """
+        df = self.trace.df_event('pixel6_emeter')
+        channels = list(Pixel6Analysis.EMETER_CHAN_NAMES.keys())
+        df = df.groupby(['chan_name'], observed=True).apply(
+            lambda x: df_add_delta(x, col='value_diff', src_col='value', window=self.trace.window)['value_diff'] / df_add_delta(x, col='ts_diff', src_col='ts', window=self.trace.window)['ts_diff']
+        ).reset_index().query("chan_name in @channels").rename(columns={0: 'value', 'chan_name': 'channel'}).dropna().set_index('Time')
+        df['channel'] = df['channel'].apply(lambda s: Pixel6Analysis.EMETER_CHAN_NAMES[s])
+        return df
+
+###############################################################################
+# Plotting methods
+###############################################################################
+
+    @TraceAnalysisBase.plot_method
+    @df_power_meter.used_events
+    def plot_power_meter(self, channels=None):
+        """
+        Plot the power meter readings from various channels.
+
+        :param channels: List of channels to plot
+        :type channels: list(str)
+
+        The channels needs to correspond to values in the ``channel`` column of df_power_meter().
+        """
+        df = self.df_power_meter()
+
+        channels = channels if channels is not None else list(df['channel'].unique())
+        if any(channel not in df['channel'].unique() for channel in channels):
+            raise ValueError('Specified channel not found')
+
+        channel_data = {channel: data for channel, data in df.groupby(['channel'], observed=True)}
```

Comment:
That doesn't work. The return of groupby is a list of tuples `(channel_name, df)`. It can be iterated over like above or I can turn it into a list but trying dict just errors out with `TypeError: attribute of type 'list' is not callable`

Reply:
Of course it does not ...
https://github.com/pandas-dev/pandas/issues/30621
So turns out that:
1. `dict()` treats as a mapping anything that has `.keys()` and `__getitem__` and will use those methods.
2. treats everything else as a key/value iterator.

Pandas' groupby object falls in category 1. except that `.keys()` just raises random errors depending on circumstances (your TypeError). 

So we can dodge it by hiding the `.keys()` behind a layer of `iter()`: `dict(iter(df.groupby(['channel'], observed=True)))`